File: Python/rotate-list.py
# Definition for singly-linked list.

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        # Time:  O(n)
        # Space: O(1)
        if head is None or head.next is None:
            return head

        size = 1
        cur = head
        while cur.next is not None:
            size += 1
            cur = cur.next
        cur.next = head

        k = k % size

        tail = cur
        cur = head
        for _ in range(size - k):
            tail = cur
            cur = cur.next
        tail.next = None
        return cur

File: C++/rotate-list.cpp
// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};


class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        // Time:  O(n)
        // Space: O(1)
        if (head == nullptr || head->next == nullptr ) {
            return head;
        }

        int size = 1;
        ListNode* cur = head;
        while (cur->next) {
            size++;
            cur = cur->next;
        }
        cur->next = head;

        k %= size;

        ListNode* tail = cur;
        cur = head;
        for (int i = 0; i < size - k; cur = cur->next, i++) {
            tail = cur;
        }
        tail->next = nullptr;
        return cur;
    }
};
